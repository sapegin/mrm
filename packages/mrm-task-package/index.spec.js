jest.mock('fs');
jest.mock('git-username');
jest.mock('mrm-core/src/util/log', () => ({
	added: jest.fn(),
}));

const path = require('path');
const { getTaskOptions } = require('mrm');
const vol = require('memfs').vol;
const task = require('./index');

const options = {
	name: 'Gendalf',
	url: 'https://middleearth.com',
	github: 'gendalf',
};

afterEach(() => {
	vol.reset();
	process.chdir('/');
});

it('should throw when bad `name`', () => {
	expect(
		getTaskOptions(task, false, {
			name: '',
		})
	).rejects.toThrow('Missing required config options: name.');
});

it('should throw when bad `github`', () => {
	expect(
		getTaskOptions(task, false, {
			github: '',
		})
	).rejects.toThrow('Missing required config options: github.');
});

it('should add package.json', async () => {
	// The task will use the folder name as a package name
	vol.mkdirpSync(__dirname);
	process.chdir(__dirname);

	task(await getTaskOptions(task, false, options));

	expect(vol.toJSON()[path.join(__dirname, 'package.json')]).toMatchSnapshot();
});

it('should add package-lock.json if packageManager is npm', async () => {
	// The task will use the folder name as a package name
	vol.mkdirpSync(__dirname);
	process.chdir(__dirname);

	task(
		await getTaskOptions(
			task,
			false,
			Object.assign({}, options, {
				packageManager: 'npm',
			})
		)
	);

	expect(vol.toJSON()[path.join(__dirname, 'package-lock.json')]).toBeDefined();
});

it('should not add package-lock.json if npmrc has false for package-lock.json', async () => {
	vol.fromJSON({
		[`${__dirname}/.npmrc`]: 'package-lock=false',
	});

	// The task will use the folder name as a package name
	vol.mkdirpSync(__dirname);
	process.chdir(__dirname);

	task(
		await getTaskOptions(
			task,
			false,
			Object.assign({}, options, {
				packageManager: 'npm',
			})
		)
	);

	expect(
		vol.toJSON()[path.join(__dirname, 'package-lock.json')]
	).not.toBeDefined();
});

it('should add package-lock.json if npmrc has true for package-lock.json', async () => {
	vol.fromJSON({
		[`${__dirname}/.npmrc`]: 'package-lock=true',
	});

	// The task will use the folder name as a package name
	vol.mkdirpSync(__dirname);
	process.chdir(__dirname);

	task(
		await getTaskOptions(
			task,
			false,
			Object.assign({}, options, {
				packageManager: 'npm',
			})
		)
	);

	expect(vol.toJSON()[path.join(__dirname, 'package-lock.json')]).toBeDefined();
});

it('should add yarn.lock if packageManager is Yarn', async () => {
	// The task will use the folder name as a package name
	vol.mkdirpSync(__dirname);
	process.chdir(__dirname);

	task(
		await getTaskOptions(
			task,
			false,
			Object.assign({}, options, {
				packageManager: 'Yarn',
			})
		)
	);

	expect(vol.toJSON()[path.join(__dirname, 'yarn.lock')]).toBeDefined();
});

it('should add .yarnrc.yml if packageManager is Yarn Berry', async () => {
	// The task will use the folder name as a package name
	vol.mkdirpSync(__dirname);
	process.chdir(__dirname);

	task(
		await getTaskOptions(
			task,
			false,
			Object.assign({}, options, {
				packageManager: 'Yarn Berry',
			})
		)
	);

	expect(vol.toJSON()[path.join(__dirname, '.yarnrc.yml')]).toBeDefined();
});

it('should add pnpm-lock.yaml if packageManager is pnpm', async () => {
	// The task will use the folder name as a package name
	vol.mkdirpSync(__dirname);
	process.chdir(__dirname);

	task(
		await getTaskOptions(
			task,
			false,
			Object.assign({}, options, {
				packageManager: 'pnpm',
			})
		)
	);

	expect(vol.toJSON()[path.join(__dirname, 'pnpm-lock.yaml')]).toBeDefined();
});

it('should set custom Node.js version', async () => {
	vol.fromJSON({
		[`${__dirname}/package-lock.json`]: '',
	});
	task(
		await getTaskOptions(
			task,
			false,
			Object.assign({}, options, {
				minNode: '9.1',
			})
		)
	);
	expect(vol.toJSON()['/package.json']).toMatchSnapshot();

	// Ok with already-existing
	task(
		await getTaskOptions(
			task,
			false,
			Object.assign({}, options, {
				minNode: '9.1',
			})
		)
	);
	expect(vol.toJSON()['/package.json']).toMatchSnapshot();
});

it('should set custom license', async () => {
	vol.fromJSON({
		[`${__dirname}/package-lock.json`]: '',
	});
	task(
		await getTaskOptions(
			task,
			false,
			Object.assign({}, options, {
				license: 'BSD',
			})
		)
	);
	expect(vol.toJSON()['/package.json']).toMatchSnapshot();
});
