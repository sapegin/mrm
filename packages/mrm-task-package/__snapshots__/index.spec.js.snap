// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should add package.json 1`] = `
"{
  \\"name\\": \\"mrm-task-package\\",
  \\"version\\": \\"1.0.0\\",
  \\"description\\": \\"\\",
  \\"author\\": {
    \\"name\\": \\"Gendalf\\",
    \\"url\\": \\"https://middleearth.com\\"
  },
  \\"contributors\\": [],
  \\"homepage\\": \\"https://github.com/gendalf/mrm-task-package\\",
  \\"bugs\\": \\"https://github.com/gendalf/mrm-task-package/issues\\",
  \\"repository\\": \\"https://github.com/gendalf/mrm-task-package\\",
  \\"license\\": \\"MIT\\",
  \\"engines\\": {
    \\"node\\": \\">=10\\"
  },
  \\"main\\": \\"index.js\\",
  \\"files\\": [
    \\"index.js\\"
  ],
  \\"scripts\\": {},
  \\"keywords\\": [],
  \\"dependencies\\": {},
  \\"devDependencies\\": {}
}"
`;

exports[`should set custom Node.js version 1`] = `
"{
  \\"name\\": \\"\\",
  \\"version\\": \\"1.0.0\\",
  \\"description\\": \\"\\",
  \\"author\\": {
    \\"name\\": \\"Gendalf\\",
    \\"url\\": \\"https://middleearth.com\\"
  },
  \\"contributors\\": [],
  \\"homepage\\": \\"https://github.com/gendalf/\\",
  \\"bugs\\": \\"https://github.com/gendalf//issues\\",
  \\"repository\\": \\"https://github.com/gendalf/\\",
  \\"license\\": \\"MIT\\",
  \\"engines\\": {
    \\"node\\": \\">=9.1\\"
  },
  \\"main\\": \\"index.js\\",
  \\"files\\": [
    \\"index.js\\"
  ],
  \\"scripts\\": {},
  \\"keywords\\": [],
  \\"dependencies\\": {},
  \\"devDependencies\\": {}
}"
`;

exports[`should set custom license 1`] = `
"{
  \\"name\\": \\"\\",
  \\"version\\": \\"1.0.0\\",
  \\"description\\": \\"\\",
  \\"author\\": {
    \\"name\\": \\"Gendalf\\",
    \\"url\\": \\"https://middleearth.com\\"
  },
  \\"contributors\\": [],
  \\"homepage\\": \\"https://github.com/gendalf/\\",
  \\"bugs\\": \\"https://github.com/gendalf//issues\\",
  \\"repository\\": \\"https://github.com/gendalf/\\",
  \\"license\\": \\"BSD\\",
  \\"engines\\": {
    \\"node\\": \\">=10\\"
  },
  \\"main\\": \\"index.js\\",
  \\"files\\": [
    \\"index.js\\"
  ],
  \\"scripts\\": {},
  \\"keywords\\": [],
  \\"dependencies\\": {},
  \\"devDependencies\\": {}
}"
`;
